drivers/syslog/syslog_channel.c

```diff
@@ -141,8 +145,14 @@ int syslog_channel(FAR const struct syslog_channel_s *channel)
     {
       DEBUGASSERT(channel->sc_putc != NULL && channel->sc_force != NULL);
 
-      g_syslog_channel = channel;
-      return OK;
+      for (int i = 0; i < CONFIG_SYSLOG_MAX_CHANNELS; i++)
```

Comment:
Should we keep the old semantics(replace the default one)? and add a new function(e.g. syslog_channel_add or syslog_channel_register).

Reply:
I don't think this will work as intended.  
`syslog_channel` is used in many places, like `syslog_file_channel` or `syslog_console_channel`.

I would expect a call to `syslog_file_channel` for example to add another channel, instead of overwriting the old one.

Alternatively, `syslog_channel` may be completely removed, and substituted with `syslog_channel_add` and `syslog_channel_remove`, so the user has the ability to set up whatever log channels combination they want?

Reply:
I am thinking that a function like `syslog_channel_remove` will be needed.  
For example, in cases like logging in a file in an external medium (e.g. SD Card), the syslog channel must be removed when the medium is also unmounted.

In the case of adding and removing channels dynamically, what will be the default channel?  
It will not be sure that it will be the first in the list...

I am leaning more towards dropping `syslog_channel` and adding `syslog_channel_add` & `syslog_channel_remove`, if you agree.

Reply:
The problem by changing `syslog_channel` from replacement to addition will break the usage in the default config(CONFIG_SYSLOG_MAX_CHANNELS == 1), because syslog_channel always fail to find the empty slot. We should at leaset keep the replacement semantics for CONFIG_SYSLOG_MAX_CHANNELS == 1.

Reply:
And what is the "default" channel that `syslog_channel` will substitute?

Functions like `syslog_file_channel` shall call `syslog_channel` or `syslog_channel_add`?

Reply:
Maybe, we can change syslog_channel behaviour base on the value of CONFIG_SYSLOG_MAX_CHANNELS:

1. Use the replacement semantics when CONFIG_SYSLOG_MAX_CHANNELS == 1
2. Use the addition semantics when CONFIG_SYSLOG_MAX_CHANNELS > 1

At least, we should allow user switch to syslog_file_channel when CONFIG_SYSLOG_MAX_CHANNELS equals 1

Reply:
I think this will be very confusing for the users... I don't like dual functionality of a single function.

No matter what is the final interface, the operation of each function should be clear and constant.

Reply:
I can think of another solution. (Which I also don't like very much...)

Maybe the first call to `syslog_channel` will substitute the default channel, and every subsequent call will append the new channel to the list.

On the other hand, any reason of not just moving to `_add` & `_remove` functions?  
I will most probably add the `_remove` variant in this PR, because it is needed (I just don't have the time today).

Reply:
> I can think of another solution. (Which I also don't like very much...)
> 
> Maybe the first call to `syslog_channel` will substitute the default channel, and every subsequent call will append the new channel to the list.
> 
> On the other hand, any reason of not just moving to `_add` & `_remove` functions?

`_add` functions can't resolve the problem I mentioned before. We need the replacement semantics when CONFIG_SYSLOG_MAX_CHANNELS == 1 regardless how you name the function.

> I will most probably add the `_remove` variant in this PR, because it is needed (I just don't have the time today).

Yes, I think so.

Reply:
OK, lets recap this:

- `syslog_channel` will be retained and it will substitute the default channel with the provided one.
- `syslog_channel_add` will be introduced and it will append a new channel to the list.
- `syslog_channel_remove` will be introduced and it will remove a channel from the list.

Some questions:

- Which is the "default" channel? May I assume that it will always be the first in the list?
- Functions like `syslog_file_channel` will use the substitute or the add function? I would go for the addition.

Reply:
> OK, lets recap this:
> 
> * `syslog_channel` will be retained and it will substitute the default channel with the provided one.
> * `syslog_channel_add` will be introduced and it will append a new channel to the list.
> * `syslog_channel_remove` will be introduced and it will remove a channel from the list.
> 
> Some questions:
> 
> * Which is the "default" channel? May I assume that it will always be the first in the list?

Most likely the first, or if `syslog_channel` remove all channel before add self then we don't need to know which one is the default.

> * Functions like `syslog_file_channel` will use the substitute or the add function? I would go for the addition.

But, it will fail to add in case of CONFIG_SYSLOG_MAX_CHANNELS == 1. That's why I suggest that `syslog_channel` change the behaviour based on CONFIG_SYSLOG_MAX_CHANNELS, or syslog_file_channel call different API(syslog_channel/syslog_channel_add) to register self.

Reply:
I added the `_remove` function.

Please give me some time to think about the optimal implementation of `syslog_channel`, regarding appending / overwritting etc...

Reply:
Sure.

Reply:
@xiaoxiang781216 I settled in for the implementation that you propose.  
Please review, so I can proceed with documentation.